### PR TITLE
Boomerang

### PR DESCRIPTION
**Pull request by ctubio**
Hello michael''
here is a new quoting mode like PingPong but taking into account not only the last trade but also all other previous trades still opened.

A trade is considered opened once it is done, and will remain as opened until another matching trade of the opposite side plus the margin of the width is also done. Then both trades are merged and flagged as closed.

Closed trades as well as the whole history of trades can be cleaned using ui buttons.
Also the ui features optional audio and the size input was split into two inputs: buySz and sellSz.
Message table is not displayed anymore and market_display table took over its space.

Included in this PR, is also #74 and #75, and a major silent of console.log and filesystem log messages (together with truncated database collections to 1 single record [less than trades collection, that is unlimited and is loaded all at startup without limit by start_time]) to avoid hitting the limit of the harddisk after 2 hours of run.

Sorry for the horrible commit messages, i didnt expected for this to be shared, but here it comes!

### CODE REVIEW

File: src/service/main.ts [hunk 1]
```diff
@@ -106,34 +106,33 @@ var messagingLog = Utils.log("tribeca:messaging");
...
-    
+
     return new Models.CurrencyPair(Models.Currency[split[0]], Models.Currency[split[1]]);
 }
 var pair = ParseCurrencyPair(config.GetString("TradedPair"));
 
 var defaultActive : Models.SerializedQuotesActive = new Models.SerializedQuotesActive(false, moment.unix(1));
-var defaultQuotingParameters : Models.QuotingParameters = new Models.QuotingParameters(.3, .05, Models.QuotingMode.Top, 
-    Models.FairValueModel.BBO, 3, .8, false, Models.AutoPositionMode.Off, false, 2.5, 300, .095, 2*.095, .095, 3, .1);
+var defaultQuotingParameters : Models.QuotingParameters = new Models.QuotingParameters(2, 0.02, 0.01, Models.QuotingMode.Boomerang, Models.FairValueModel.BBO, 1, 0.9, true, Models.AutoPositionMode.EwmaBasic, false, 0.9, 569, false, .095, 2*.095, .095, 3, .1);
```
Comment (ctubio) [inline src/service/main.ts]: im sorry i had deffined new defaults, you can set your own back as you wish ofcourse

File: src/admin/trades.ts [hunk 2]
```diff
@@ -69,24 +76,79 @@ var TradesListController = ($scope : TradesScope, $log : ng.ILogService, subscri
...
     $scope.$on('$destroy', () => {
         sub.disconnect();
-        $log.info("destroy trades list");
+        qpSub.disconnect();
+        // $log.info("destroy trades list");
     });
 
-    $log.info("started trades list");
+    // $log.info("started trades list");
+    setTimeout(function(){$scope.audioReady = true;},7000);
```
Comment (ctubio) [inline src/admin/trades.ts]: this timer is to avoid the sound when loading trades on strartup

File: src/common/models.ts [hunk 3]
```diff
@@ -58,7 +58,7 @@ export class MarketTrade implements ITimestamped {
 }
 
 export enum GatewayType { MarketData, OrderEntry, Position }
-export enum Currency { USD, BTC, LTC, EUR, GBP, CNY , ETH }
+export enum Currency { USD, BTC, LTC, EUR, GBP, CNY , ETH , CAD }
```
Comment (ctubio) [inline src/common/models.ts]: required by the new coinbase sandbox api

**Comment by Camille92** [comment]
You were fast to pull the merge request :)

File: src/admin/trade-safety.ts [hunk 4]
```diff
@@ -38,18 +44,18 @@ var TradeSafetyController = ($scope : TradeSafetyScope, $log : ng.ILogService, s
...
+    // $log.info("started trade safety");
 };
 
 export var tradeSafetyDirective = "tradeSafetyDirective";
 
 angular
     .module(tradeSafetyDirective, ['sharedDirectives'])
     .directive("tradeSafety", () => {
-        var template = '<span>BuyTS: {{ buySafety|number:2 }}, SellTS: {{ sellSafety|number:2 }}, TotalTS: {{ tradeSafetyValue|number:2 }}</span>';
+        var template = '<span>BuyLT: <b>{{ buySizeSafety|number:2 }}</b>, SellLT: <b>{{ sellSizeSafety|number:2 }}</b>, BuyTS: {{ buySafety|number:2 }}, SellTS: {{ sellSafety|number:2 }}, TotalTS: {{ tradeSafetyValue|number:2 }}</span>';
```
Comment (ctubio) [inline src/admin/trade-safety.ts]: this displays the current safety values based on the history of trades

File: src/service/broker.ts [hunk 5]
```diff
@@ -168,6 +233,50 @@ export class OrderBroker implements Interfaces.IOrderBroker {
         this.onOrderUpdate(rpt);
     };
 
+    private _reTrade = (reTrades: Models.Trade[], trade: Models.Trade) => {
```
Comment (ctubio) [inline src/service/broker.ts]: founds and closes matching trades

File: src/service/persister.ts [hunk 6]
```diff
@@ -162,13 +168,59 @@ export class Persister<T extends Persistable> implements ILoadAll<T> {
     public persist = (report: T) => {
         this.collection.then(coll => {
             this._saver(report);
+            if (["fv","md","msg","mt","pos","qp-sub","tbp","osr","tsv"].indexOf(this._dbName)>-1)
+              coll.deleteMany({ time: { $exists:true } }, err => {
+                  if (err)
+                      this._log.error(err, "Unable to deleteMany", this._dbName, report);
+              });
```
Comment (ctubio) [inline src/service/persister.ts]: truncate collections to 1 single record to save space in very very long runs; only trades table persist unlimited records

File: src/service/persister.ts [hunk 7]
```diff
@@ -162,13 +168,59 @@ export class Persister<T extends Persistable> implements ILoadAll<T> {
...
+                    _.forEach(arr, this._loader);
+                    deferred.resolve(arr);
+                }
+            });;
+        }).done();
+
+        return deferred.promise;
+    };
+
+    public repersist = (report: T, trade: Models.Trade) => {
```
Comment (ctubio) [inline src/service/persister.ts]: updates matching trades and deletes closed trades

File: src/service/quoting-engine.ts [hunk 8]
```diff
@@ -94,37 +94,52 @@ export class QuotingEngine {
...
+            }
+          }
+        if (unrounded.bidPx !== null)
+          for (var fbi = 0; fbi < filteredMkt.bids.length; fbi++) {
+            if (filteredMkt.bids[fbi].price < unrounded.bidPx) {
+              unrounded.bidPx = filteredMkt.bids[fbi].price + .01;
+              break;
+            }
+          }
+
```
Comment (ctubio) [inline src/service/quoting-engine.ts]: price accomodation to avoid "hollows" in the book and attempt to get "as much as possible"

File: src/static/index.html [hunk 9]
```diff
@@ -53,7 +53,7 @@
             <td>mkt{{ $index }}</td>
             <td>{{ level.bidSize|number:2 }}</td>
             <td ng-class="level.bidClass">{{ level.bidPrice|number:2 }}</td>
-            <td></td>
+            <td><span ng-show="level.diffWidth > 0">{{ level.diffWidth|number:2 }}</span></td>
```
Comment (ctubio) [inline src/static/index.html]: show diff betweeen current trades in both sides

File: src/service/quoting-engine.ts (hunk 8 above)
Comment (Camille92) [inline src/service/quoting-engine.ts]: This is a real plus !

File: src/service/broker.ts [hunk 10]
```diff
@@ -168,6 +233,50 @@ export class OrderBroker implements Interfaces.IOrderBroker {
...
+          if (this._trades[i].tradeId==reTrade.tradeId) {
+            gowhile = true;
+            var allocQty = Math.min(trade.quantity, this._trades[i].quantity - this._trades[i].alloc);
+            this._trades[i].time = trade.time;
+            this._trades[i].allocprice = ((allocQty*trade.price) + (this._trades[i].alloc*this._trades[i].allocprice)) / (this._trades[i].alloc+allocQty);
+            this._trades[i].alloc += allocQty;
+            trade.quantity -= allocQty;
+            trade.value = Math.abs(trade.price*trade.quantity);
+            if (this._trades[i].quantity<=this._trades[i].alloc)
+              this._trades[i].value = Math.abs((this._trades[i].quantity*this._trades[i].price)-(this._trades[i].alloc*this._trades[i].allocprice));
```
Comment (ctubio) [inline src/service/broker.ts]: display the total income earned after a trade is closed

File: src/service/gateways/coinbase.ts [hunk 11]
```diff
@@ -395,11 +395,11 @@ class CoinbaseMarketDataGateway implements Interfaces.IMarketDataGateway {
     private _cachedAsks: Models.MarketSide[] = null;
 
     private reevalBids = () => {
-        this._cachedBids = _.map(this._orderBook.bids.store.slice(0, 3), s => (<any>s).value.marketUpdate);
+        this._cachedBids = _.map(this._orderBook.bids.store.slice(0, 9), s => (<any>s).value.marketUpdate);
     };
 
     private reevalAsks = () => {
-        this._cachedAsks = _.map(this._orderBook.asks.store.slice(0, 3), s => (<any>s).value.marketUpdate);
+        this._cachedAsks = _.map(this._orderBook.asks.store.slice(0, 9), s => (<any>s).value.marketUpdate);
```
Comment (ctubio) [inline src/service/gateways/coinbase.ts]: displays more data of the book in the frontend (since message table is hidden) and also it allows more options for calculate a price accomodation (applied to all quoting modes)

File: src/service/broker.ts [hunk 12]
```diff
@@ -168,6 +233,50 @@ export class OrderBroker implements Interfaces.IOrderBroker {
...
+            this._tradePersister.repersist(this._trades[i], this._trades[i]);
+            break;
+          }
+        }
+      }
+      this.Trade.trigger(trade);
+      if (trade.quantity>0) {
+        var exists = false;
+        for(var i = 0;i<this._trades.length;i++) {
+          if (this._trades[i].price==trade.price && this._trades[i].side==trade.side && this._trades[i].quantity>this._trades[i].alloc) {
```
Comment (ctubio) [inline src/service/broker.ts]: accumulate opened trades of the same prices and side

File: src/admin/client.ts [hunk 13]
```diff
@@ -74,7 +76,7 @@ class DisplayOrder {
 
     public submit = () => {
         var msg = new Models.OrderRequestFromUI(this.side, this.price, this.quantity, this.timeInForce, this.orderType);
-        this._log.info("submitting order", msg);
```
Comment (michaelgrosner) [inline src/admin/client.ts]: Why are you getting rid of the client-side logging?

File: src/service/gateways/coinbase.ts (hunk 11 above)
Comment (michaelgrosner) [inline src/service/gateways/coinbase.ts]: I'd like this to eventually be either externally configurable or something because I've found there to be value with more depth shown, but for now this is fine.

File: src/service/main.ts (hunk 1 above)
Comment (michaelgrosner) [inline src/service/main.ts]: This is fine.

File: src/service/main.ts [hunk 14]
```diff
@@ -406,19 +409,18 @@ var runTradingSystem = (classes: SimulationClasses) : Q.Promise<boolean> => {
...
         // event looped blocked timer
         var start = process.hrtime();
         var interval = 100;
         setInterval(() => {
             var delta = process.hrtime(start);
             var ms = (delta[0] * 1e9 + delta[1]) / 1e6;
             var n = ms - interval;
-            if (n > 25)
-                mainLog.info("Event looped blocked for " + Utils.roundFloat(n) + "ms");
+            ////if (n > 25) mainLog.info("Event looped blocked for " + Utils.roundFloat(n) + "ms");
```
Comment (michaelgrosner) [inline src/service/main.ts]: Please do not remove this logging. It's quite useful to determine if there is stuff clogging up the event loop.

File: src/admin/trades.ts (hunk 2 above)
Comment (michaelgrosner) [inline src/admin/trades.ts]: One idea is to affix a is-loaded-from-the-database-on-startup flag to all the Trade objects, then not triggering the sound on those trades. I could see a flag like that being useful in other contexts.

File: src/service/persister.ts (hunk 7 above)
Comment (michaelgrosner) [inline src/service/persister.ts]: I am very against these changes to persister.ts. The data, once persisted, should never change and data in the database should always be immutable. This is good from a code-quality standpoint (reduces cohesion between persister and the quoting logic) but also from a data analysis standpoint (I can recreate the state of the world at any time solely from whats in the DB while performing my analysis)

If there needs to be calculations done to the persisted data, the code should fetch it from the DB, calculate the new value and use it if needed, then persist it back out into a new table/collection if it's even needed.

The interface of IPersist<T>/ILoad*<T> should not need to know about the implementations of various quoting styles or really even what data is being flung at them. The addition of trade and side to the interface violates that.

In this case, I would recommend the approach of having a second collection of closed trade IDs, having a secondary component figure out which trades are opened and closed, then persist out what's actually closed.

File: src/admin/client.ts [hunk 15]
```diff
@@ -84,15 +86,21 @@ var uiCtrl = ($scope : MainWindowScope,
               $log : ng.ILogService,
               subscriberFactory : Shared.SubscriberFactory,
               fireFactory : Shared.FireFactory) => {
-    
+
     var cancelAllFirer = fireFactory.getFire(Messaging.Topics.CancelAllOrders);
     $scope.cancelAllOrders = () => cancelAllFirer.fire(new Models.CancelAllOrdersRequest());
-                  
+
+    var cleanAllClosedFirer = fireFactory.getFire(Messaging.Topics.CleanAllClosedOrders);
```
Comment (michaelgrosner) [inline src/admin/client.ts]: Can you explain the trading necessity to have a UI element for cleaning trades? Is cleaning a manual process when the market moves away enough that trading won't match any longer?

File: src/service/broker.ts (hunk 10 above)
Comment (michaelgrosner) [inline src/service/broker.ts]: Similar to the discussion of IPersist, this code seems fairly out of place and blends quoting logic-specific code in a place that shouldn't really care about quoting mode.

There are several IOrderEntryGateways providing exchange-specific implementations of order entry but translates their data into a common API. The OrderBroker's single responsibility is to provide a stateful abstraction around an IOrderEntryGateway, so users of order entry can query state, listen to order updates, and submit orders/cancels. It should be entirely possible to take the OrderBroker implementation as is and drop it into an arbitrage bot or TA bot (I've actually done this several times while testing new trading styles). This logic doesn't belong at this level and should be at a level above the OrderBroker, abstracted out.

File: src/admin/client.ts (hunk 13 above)
Comment (ctubio) [inline src/admin/client.ts]: this is not related to the full space of the vmdk of docker, but instead is to minimize the traffic of the application. I preffer the app to send data to the exchange api rathen than to the client (if that msg to the client is just a notice of an action already known by the client by other means [rich gui]).

Also, users do not look at the console; if this is for devs, let devs activate (uncomment) this stream of debuggin data when needed rather than force all users to transmit it..

if you dont agree i can uncomment it in a next commit easy  :blush:

File: src/service/main.ts (hunk 14 above)
Comment (ctubio) [inline src/service/main.ts]: im sorry, fixed at https://github.com/michaelgrosner/tribeca/pull/76/commits/71a2625b3a71ec6f012e0a18938e4e610a3f3a9d

File: src/admin/client.ts (hunk 13 above)
Comment (michaelgrosner) [inline src/admin/client.ts]: I'm not following the first paragraph "this is not related to the full space of the vmdk of docker...".

The logging is for development but if you'd like the ability to eliminate it, I'd prefer either

a) Changing it to Debug level then using logProvider (https://docs.angularjs.org/api/ng/provider/$logProvider) to turn it off
a) Wholly remove it, don't leave the commented version around.

File: src/admin/trades.ts (hunk 2 above)
Comment (ctubio) [inline src/admin/trades.ts]: thanks! marked as TODO soon

File: src/service/persister.ts (hunk 7 above)
Comment (ctubio) [inline src/service/persister.ts]: thanks! marked as TODO soon

File: src/service/broker.ts (hunk 10 above)
Comment (ctubio) [inline src/service/broker.ts]: thanks! marked as TODO soon

File: src/admin/client.ts (hunk 13 above)
Comment (ctubio) [inline src/admin/client.ts]: thanks! marked as TODO soon (will try option A)

about the vmdk of docker comment, i found the hard disk of my virtual boxes completely full after a few days of run.

File: src/admin/client.ts (hunk 15 above)
Comment (ctubio) [inline src/admin/client.ts]: I added this button because now trades are always displayed forever. The user can look at all the trades opened and also at all closed trades where the final net income is displayed.

Using this button, the user can manualy clean the closed trades and "start a new cycle" anytime he/she wishes.

For example, for myself, i wakeup on the morning and see all closed trades during the night with all the net incomes. Then i clean all closed trades and go to work. Once im back in the afternoon i can examinate what new trades has been closed and the net incomes during the day. Then clean them and go to sleep. And so on every day :D

The other button of clean all orders, is because what you said, in case the market moves way too off, you can manualy clean all the historic and therefore assume the losses if any.

**Comment by Camille92** [comment]
(BUG) Hi guys,

Just to let you know that trades on OkCoin using the last version of the Boomerang branch have 0 persistence (trade are posted and directly canceled within 1ms).

I don't know since when it is happening but versions prior to 25th of July doesn't have the issue.

File: src/admin/client.ts (hunk 15 above)
Comment (obsidian0001) [inline src/admin/client.ts]: @ctubio Are you running these current changes in this pull request? Is everything stable?
